llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+				if (const ConstantInt* CI = dyn_cast<ConstantInt>(I.getOperand(1)))
+				{
+					if (CI->getZExtValue() <= min)
+						return BitMask::ALL;
+				}
+
+				[[clang::fallthrough]];
+			}
+			case Instruction::Or:
+			case Instruction::Xor:
+			{
+				//Those all work modulo 2^k
+				return min;
+			}
+			default:
+			{
+				//Some operands NOT fully known, means in general no information is conserved
+				return BitMask::NONE;
+			}
+		}
+	}
+	void setIncomingBB(const llvm::BasicBlock* from)
+	{
+		fromBB = from;
+	}
+	const llvm::BasicBlock* getIncomingBB() const
+	{
+		return fromBB;
+	}
+	enum VisitingPolicy {COMPUTE_VALUES, REMOVE_VALUES};
+	template <VisitingPolicy policy>
+	const llvm::BasicBlock* visitBasicBlock(const llvm::BasicBlock* BB, const llvm::BasicBlock* from=nullptr);
+	explicit PartialInterpreter(std::unique_ptr<llvm::Module> M)
+		: llvm::Interpreter(std::move(M), /*preExecute*/false)
+	{
+	}
+	BitMask getBitMask(const llvm::Value* V) const
+	{
+		if (isa<ConstantData>(V))
+			return BitMask::ALL;
+
+		if (isa<Function>(V))
+			return BitMask::NONE;
+
+		if (const GlobalVariable* GV = dyn_cast<GlobalVariable>(V))
+		{
+			if (getDataLayout().getPreferredAlign(GV) == 8)
+				return BitMask::ALIGNED8;
+			if (getDataLayout().getPreferredAlign(GV) == 4)
+				return BitMask::ALIGNED4;
+			if (getDataLayout().getPreferredAlign(GV) == 2)
+				return BitMask::ALIGNED2;
+		}
+		if (const ConstantExpr* CE = dyn_cast<ConstantExpr>(V))
+		{
+			if (CE->getOpcode() == Instruction::GetElementPtr)
+			{
+				const GEPOperator *GEP = cast<GEPOperator>(CE);
+				if (GEP->hasAllZeroIndices())
+					return getBitMask(CE->getOperand(0));
+			}
+		}
+
+		auto it = stronglyKnownBits.find(V);
+		if (it != stronglyKnownBits.end())
+			return it->second;
+
+		return BitMask::NONE;
+	}
+	bool isValueComputed(const llvm::Value* V) const
+	{
+		if (isValueComputedConstant(V))
+			return true;
+		if (isa<BasicBlock>(V))
+			return true;
+		if (isa<Argument>(V) || isa<Instruction>(V))
+		{
+			return stronglyKnownBits.count(V);
+		}
+		return false;
+	}
+	bool areOperandsComputed(const llvm::Instruction& I)
+	{
+		if (isa<SelectInst>(I))
+		{
+			if (isValueComputed(I.getOperand(0)))
+			{
+				GenericValue V = getOperandValue(I.getOperand(0));
+				if (V.IntVal == 0u)
+				{
+					if (isValueComputed(I.getOperand(2)))
+						return true;
+				}
+				else
+				{
+					if (isValueComputed(I.getOperand(1)))
+						return true;
+				}
+
+			}	
+		}
+		for (auto& op : I.operands())
+		{
+			if (!isValueComputed(op))
+				return false;
+		}
+		if (const SwitchInst* SI = dyn_cast <SwitchInst>(&I))
+		{
+			if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
+				return false;
+		}
+		if (const BranchInst* SI = dyn_cast <BranchInst>(&I))
+		{
+			if (SI->isConditional())
+				if ((getBitMask(SI->getCondition()) != BitMask::ALL) )
```

Comment:
The condition is i1, so there is no point in checking the bitmask.

Reply:
getBitMask will return NONE in some patological cases (actually == NONE it's clearer)